rabbitsmpp/publisher.go

```diff
@@ -2,71 +2,271 @@ package rabbitsmpp
+
+	config   Config
+	ampqConn *amqp.Connection
+	bound    bool
+
+	closed bool
+}
+
+func newPublisherClient(conf Config) (*publisherClient, error) {
+	var m sync.Mutex
+
+	return &publisherClient{
+		sm:     m,
+		config: conf,
+	}, nil
+}
+
+func (c *publisherClient) Bind() (chan *amqp.Error, error) {
+	c.sm.Lock()
+	defer c.sm.Unlock()
+
+	if c.bound {
+		return nil, errAlreadyBound
+	}
+
+	if c.closed {
+		return nil, errClientClosed
+	}
+
+	conn, err := amqp.Dial(c.config.URL)
+	if err != nil {
+		return nil, err
+	}
+
+	errors := make(chan *amqp.Error)
+	errChan := conn.NotifyClose(errors)
+	c.ampqConn = conn
+
+	return errChan, nil
+}
+
+func (c *publisherClient) Channel() (Channel, error) {
+	c.sm.Lock()
+	defer c.sm.Unlock()
+
+	if c.closed {
+		return nil, errClientClosed
+	}
+
+	return c.ampqConn.Channel()
+}
+
+func (c *publisherClient) ReBind() (PublisherClient, chan *amqp.Error, error) {
+	c.sm.Lock()
+	defer c.sm.Unlock()
+
+	ticker := backoff.NewTicker(backoff.NewExponentialBackOff())
+	failed := 0
+	for _ = range ticker.C {
+		failed++
+		newClient, err := newPublisherClient(c.config)
+		if err != nil {
+			continue
+		}
+
+		errChan, err := newClient.Bind()
+		if err != nil {
+			continue
+		}
+
+		// stop the ticker
+		ticker.Stop()
+		return newClient, errChan, nil
+	}
+
+	return nil, nil, errClientClosed
+
+}
+
+func (c *publisherClient) Close() error {
+	c.sm.Lock()
+	defer c.sm.Unlock()
+
+	if c.closed {
+		return nil
+	}
+
+	c.closed = true
+	c.ampqConn.Close()
+	return nil
+}
+
+func (c *publisherClient) QueueName() string {
+	c.sm.Lock()
+	defer c.sm.Unlock()
+
+	return c.config.QueueName
+}
+
+// Publisher is interface for publishing data to the queues
 type Publisher interface {
 	Publish(Job) error
+	Close() error
+}
+
+func runWithRecovery(runCmd func() error, onError func() error) error {
+	if err := runCmd(); err == nil {
+		return nil
+	}
+
+	if err := onError(); err != nil {
+		return err
+	}
+
+	err := runCmd()
+	return err
 }
 
 type publisher struct {
-	*client
+	m        sync.Mutex
+	client   PublisherClient
+	sendChan Channel
 }
 
+// NewPublisher creates a new publisher with direct exchange
 func NewPublisher(conf Config) (Publisher, error) {
-	c := NewClient(conf).(*client)
-	return &publisher{c}, nil
+	pubClient, err := newPublisherClient(conf)
+	if err != nil {
+		return nil, err
+	}
+
+	return newPublisherWithClient(pubClient)
 }
 
-func (p *publisher) Publish(j Job) error {
-	_, err := p.Bind()
-	if err != nil {
-		return err
+func newPublisherWithClient(client PublisherClient) (Publisher, error) {
+	var m sync.Mutex
+
+	publisher := &publisher{
+		m:      m,
+		client: client,
+	}
+
+	// setup all the connections here
+	if err := publisher.reset(false); err != nil {
+		return nil, err
 	}
-	defer p.conn.Close()
 
-	ch, err := p.Channel()
+	return publisher, nil
+}
+
+func (p *publisher) Close() error {
+	p.m.Lock()
+	defer p.m.Unlock()
+	p.sendChan.Close()
+	p.client.Close()
+	return nil
```

Comment:
not sure, I just try if closed ok if not it's not an error really, since when the RQ is down closing the connection fails, don't want to fail the operation